Don't do daemon-reload in the initrd

keszybz opened the pull request.

keszybz committed:
fstab-generator: add mode to check /sysroot/etc/fstab and maybe do daemon-reload

The idea is that we can peek into /sysroot/etc/fstab and figure out if there's
anything interesting there. We could use a separate binary for this, but we'd
need to duplicate most of the logic that in systemd-fstab-generator. Thus I
think it's nicer to make systemd-fstab-generator work as a multi-call binary.

If called as systemd-sysroot-fstab-check, we look for units that we'd mount and
call daemon-reload and initrd-fs.target/restart, similarly to what we did
before, but in the process itself.

keszybz committed:
units/initrd-parse-etc.service: only start units that are required

This makes use of the option switch that was added in the previous commit.
We used a pretty big hammer on a relatively small nail: we would do daemon-reload
and (in principle) allow any configuration to be changed. But in fact we only
made use of this in systemd-fstab-generator. systemd-fstab-generator filters
out all mountpoints except /usr and those marked with x-initrd.mount, i.e. on
a big majority of systems it wouldn't do anything.

Also, since systemd-fstab-generator first parses /proc/cmdline, and then
initrd's /etc/fstab, and only then /sysroot/etc/fstab, configuration in the
host would only matter if it the same mountpoint wasn't configured "earlier".
So the config in the host could be used for new mountpoints, but it couldn't
be used to amend configuration for existing mountpoints. And we wouldn't actually
remount anything, so mountpoints that were already mounted wouldn't be affected,
even if did change some config.

In the new scheme, we will parse /sysroot/etc/fstab and explicitly start
sysroot-usr.mount and other units that we just wrote. In most cases (as written
above), this will actually result in no units being created or started.

If the generator is invoked on a system with /sysroot/etc/fstab present,
behaviour is not changed and we'll create units as before. This is needed so
that if daemon-reload is later at some points, we don't "lose" those units.

There's a minor bugfix here: we honour x-initrd.mount for swaps, but we
wouldn't restart swap.target, i.e. the new swaps wouldn't necessarilly be
pulled in immediately.

keszybz committed:
man: make bootup charts narrower

docbook already indents diagrams, so there is no need to leave whitespace on
the left. Also, make the charts a bit narrower to fit better on a terminal.

keszybz committed:
booctl: do not say uuids differ if one of the uuids is unset

We allow ESP autodetection to fail, e.g. if it is not mounted, but then we'd
say that the detected one is different than the one reported by the bootloader,
which is rather confusing.

While at it, if we actually detect a mismatch, print the two uuids.

keszybz committed:
gpt-auto-generator: include device name in error

mrc0mmand commented on src/fstab-generator/fstab-generator.c:
```diff
@@ -596,7 +648,7 @@ static int parse_fstab(bool initrd) {
 
         while ((me = getmntent(f))) {
-                _cleanup_free_ char *where = NULL, *what = NULL, *canonical_where = NULL;
+                _cleanup_free_ char *where = NULL, *what = NULL, *canonical_where = NULL, *start_unit = NULL;
```
```
../build/src/fstab-generator/fstab-generator.c:651:92: error: unused variable 'start_unit' [-Werror,-Wunused-variable]
                _cleanup_free_ char *where = NULL, *what = NULL, *canonical_where = NULL, *start_unit = NULL;
```

bluca commented on src/fstab-generator/fstab-generator.c:
```diff
@@ -567,6 +577,45 @@ static int add_mount(
+                                       NULL,
+                                       "ss", *unit, "replace");
+                if (r < 0)
+                        log_error("Failed to start %s: %s", *unit, bus_error_message(&error, r));
```
`Failed to start %s, ignoring: ...`?

bluca commented on src/fstab-generator/fstab-generator.c:
```diff
@@ -1164,4 +1230,18 @@ static int run(const char *dest, const char *dest_early, const char *dest_late)
 
-DEFINE_MAIN_GENERATOR_FUNCTION(run);
+static int run(int argc, char **argv) {
+        arg_sysroot_only = streq_ptr(argv[1], "--sysroot-only");
```
shouldn't this check that if true it's running in initrd and return an error if not? There's an assert in `parse_fstab` but that's not very user friendly

bluca commented on man/bootup.xml:
```diff
@@ -225,7 +225,7 @@ emergency.service    |              |              |
     <filename>sysroot.mount</filename> unit becomes active and
     <filename>initrd-root-fs.target</filename> is reached. The service
-    <filename>initrd-parse-etc.service</filename> scans
+    <filename>initrd-sysroot-fstab.service</filename> scans
```
Is it possible that external units have deps on the old name? Should it have an alias for a while?

yuwata commented on src/fstab-generator/fstab-generator.c:
```diff
@@ -1164,4 +1230,18 @@ static int run(const char *dest, const char *dest_early, const char *dest_late)
 
-DEFINE_MAIN_GENERATOR_FUNCTION(run);
+static int run(int argc, char **argv) {
+        arg_sysroot_only = streq_ptr(argv[1], "--sysroot-only");
```
I think it is better to use an environment variable, instead of the positional argument.

keszybz replied:
I disagree. Environment variables are inherited and are useful for "global" settings and other defaults. A command-line argument is something that is clearly visible and is what should be used for something that completely changes the behaviour of a program.

(In fact, I'd say we rely on environment variables in systemd too much. E.g. their use a the sole way to enable debugging is imo hard to justify. But that's a topic for another time.)

yuwata replied:
OK. That makes sense for me.

yuwata commented on src/fstab-generator/fstab-generator.c:
```diff
@@ -567,6 +577,45 @@ static int add_mount(
+        STRV_FOREACH(unit, units_to_start) {
+                _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+
+                log_info("Running request %s/start/replace", *unit);
```
"Running request" -> "Requesting"?

keszybz replied:
This text appears in a few places and I just copied it here. I agree it should be changed, but I'll do it in a separate commit to do it everywhere.

yuwata commented on units/initrd-parse-etc.service.in:
```diff
@@ -8,17 +8,17 @@
 
 [Service]
 Type=oneshot
-ExecStartPre=-systemctl daemon-reload
```
BTW, does this line effectively run systemd-fstab-generator? What about the other generators?
IOW, running only systemd-fstab-generator is enough after this PR?

keszybz replied:
In principle, other generators could do arbitrary stuff here. But *our* generators don't: the only one that does anything is systemd-fstab-generator. (And as mentioned in the commit message, it actually doesn't do much.) So this is clearly a change in behaviour, but I think we can get away with this.

yuwata replied:
I see. Thanks.

keszybz committed:
various: reword message

yuwata commented on src/fstab-generator/fstab-generator.c:
```diff
@@ -1164,4 +1230,22 @@ static int run(const char *dest, const char *dest_early, const char *dest_late)
+
+        char **args = strv_skip(argv, 1 + arg_sysroot_only);
+
+        log_setup_generator();
```
This should be called only when running as a generator. Otherwise, `log_setup()` should be used.

keszybz replied:
Updated.

poettering commented on src/gpt-auto-generator/gpt-auto-generator.c:
```diff
@@ -53,7 +53,8 @@ static int open_parent_block_device(dev_t devnum, int *ret_fd) {
         r = sd_device_new_from_devnum(&d, 'b', devnum);
         if (r < 0)
-                return log_debug_errno(r, "Failed to open device: %m");
+                return log_debug_errno(r, "Failed to open device "DEVNUM_FORMAT_STR": %m",
```
maybe "open" is misleading actually, and "device" is imprecise.

maybe say "Failed to allocate device object for block device…"

poettering replied:
(i mean, if you touch this, let's fix his properly)

poettering commented on src/fstab-generator/fstab-generator.c:
```diff
@@ -567,6 +577,45 @@ static int add_mount(
+
+        r = bus_connect_system_systemd(&bus);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get D-Bus connection: %m");
```
this is explicitly not OK in a generator. While the generators run PID 1 blocks and just waits for them to complete. This cannot work this way, sorry. We need to find some other way.

See generator man page.

keszybz replied:
No, it's not invoked as a generator. It's effectively a multi-call binary that gets invoked as a generator or as a normal service.

poettering commented on units/initrd-parse-etc.service.in:
```diff
@@ -8,17 +8,25 @@
-ExecStart=-systemctl --no-block start initrd-fs.target
+
+ExecStart={{SYSTEM_GENERATOR_DIR}}/systemd-fstab-generator --sysroot-only \
+  /run/systemd/generator/ /run/systemd/generator.early/ /run/systemd/generator.late/
```
hmm, what? Running a generator from a service? this is supposed to add more units to the generator dirs?

that's not safe, if a unit the generator generates was already referenced before it might be loaded already and then would not be reloaded once the generator is finished.

I mean, I really dislike running generators out of generator context. This blurs the lines quite drastrically.

keszybz replied:
Please see the commit message. It explains this in some detail. (My reasoning may be wrong, but please at least read the description.)

poettering commented on src/gpt-auto-generator/gpt-auto-generator.c:
```diff
@@ -53,7 +53,8 @@ static int open_parent_block_device(dev_t devnum, int *ret_fd) {
         r = sd_device_new_from_devnum(&d, 'b', devnum);
         if (r < 0)
-                return log_debug_errno(r, "Failed to open device: %m");
+                return log_debug_errno(r, "Failed to create device object for device "DEVNUM_FORMAT_STR": %m",
```
*block* device

poettering commented on src/fstab-generator/fstab-generator.c:
```diff
@@ -570,6 +584,42 @@ static int add_mount(
+        if (r < 0)
+                return log_error_errno(r, "Failed to reload daemon: %s", bus_error_message(&error, r));
+
+        log_info("Requesting initrd-fs.target/start/replace...");
```
hmm, i think it would be good to have a comment here, explaining why we start this here. And hightlight this is about activating deps that didn't exist before the reload, but do now. This is not obvious, given that the target has already been in the queue before.

poettering commented on src/fstab-generator/fstab-generator.c:
```diff
@@ -119,6 +123,11 @@ static int add_swap(
+        if (arg_sysroot_check) {
+                log_info("%s should be enabled in the initrd, will request daemon-reload.", what);
+                return true;
+        }
```
Hmm, I think it would be a tiny bit sexier if we'd return true in this whole function even if arg_sysroot_check is not set, and we just wrote out a unit. That way the return code semantics are always the same, regardless in which mode we are invoked, i.e. regardless if we are called in the test or the generator mode.

poettering replied:
i.e. just change the `return 0` at the very end of the call to `return true;`

poettering commented on src/fstab-generator/fstab-generator.c:
```diff
@@ -378,6 +387,11 @@ static int add_mount(
+        if (arg_sysroot_check) {
+                log_info("%s should be mounted in the initrd, will request daemon-reload.", where);
+                return true;
+        }
```
similar here

poettering commented on src/fstab-generator/fstab-generator.c:
```diff
@@ -570,6 +584,42 @@ static int add_mount(
+        if (r < 0)
+                return bus_log_create_error(r);
+
+        r = sd_bus_call(bus, m, DEFAULT_TIMEOUT_USEC * 2, &error, NULL);
```
given we have at least two places now where we use the long timeout for Reload, maybe we should make it a common `#define` macro, but doesn't matter...

poettering commented on src/fstab-generator/fstab-generator.c:
```diff
@@ -1164,4 +1229,31 @@ static int run(const char *dest, const char *dest_early, const char *dest_late)
 }
 
-DEFINE_MAIN_GENERATOR_FUNCTION(run);
+static int run(int argc, char **argv) {
```
i think more descriptive would be:

```c
arg_sysrot_check = !invoked_as(argv, "systemd-fstab-generator");
if (arg_sysroot_check) {
        /* Run as in systemd-sysroot-fstab-check mode */
        …
} else {
        /* Run in generator mode */
        …
}
```

poettering replied:
or actually even maybe:

```c
arg_sysroot_check = invoked_as(argv, "systemd-sysroot-fstab-check");
…
```

keszybz committed:
Use a common define for the reload timeout

poettering commented on src/fstab-generator/fstab-generator.c:
```diff
@@ -567,6 +581,47 @@ static int add_mount(
+                                       NULL,
+                                       "ss", unit, "replace");
+                if (r < 0)
+                        return log_error_errno(r, "Failed to restart %s: %s", unit, bus_error_message(&error, r));
```
"restart" is a bit misleading, since we issue the StartUnit() call here, not RestartUnit()...

keszybz replied:
I changed that to "(re)start".

poettering commented on src/fstab-generator/fstab-generator.c:
```diff
@@ -567,6 +581,47 @@ static int add_mount(
+                                       "ss", unit, "replace");
+                if (r < 0)
+                        return log_error_errno(r, "Failed to restart %s: %s", unit, bus_error_message(&error, r));
+        }
```
hmm, and i would make error handling of the two StartUnit() calls graceful I think. but not sure.

keszybz replied:
I changed it to try both but return the first error. I think we should fail the unit if we fail to enqueue restart jobs for the targets. Normally that shouldn't happen and if it does, the job the service was doing wasn't really finished.

keszybz committed:
initrd-parse-etc: override argv[0] to avoid dracut issue

Quoting https://github.com/systemd/systemd/pull/24054#issuecomment-1210501631:
> this would need a patch in dracut, specifically adding the
> systemd-sysroot-fstab-check to the list of installed stuff:
> https://github.com/dracutdevs/dracut/blob/fe8fa2b0cadbb33e27c8dd8b5851548dcd65835c/modules.d/00systemd/module-setup.sh#L47.
>
> I could do this manually in the CI (and I guess I'd have to do it anyway even
> if the patch lands in upstream, since it won't be available in C8S), but it
> should get there first before merging this PR, otherwise it's going to break
> Rawhide.

poettering commented on src/basic/process-util.c:
```diff
@@ -1570,13 +1570,17 @@ int setpriority_closest(int priority) {
 
 bool invoked_as(char *argv[], const char *token) {
-        if (!argv || isempty(argv[0]))
+        const char *e = getenv("SYSTEMD_INVOKED_AS");
```
getenv_secure()?

also, should this maybe unset the env var, and cache the result? so that this is not propagated further down? maybe via a getenv_secure_steal() helper?

keszybz replied:
> getenv_secure()?

I considered that, but I don't see how it'd be relevant.  This is explicitly for executables, and not library code. Also, similar helpers for other variables don't use "secure_".

keszybz replied:
I dropped the commit to add $SYSTEMD_INVOKED_AS from this PR. I think it's still useful, but it needs more discussion about inheritance, etc, so let's discuss it separately.

poettering commented on units/initrd-parse-etc.service.in:
```diff
@@ -21,7 +21,10 @@ OnFailureJobMode=replace-irreversibly
+# FIXME: once dracut is patched to install the symlink, change to:
+# ExecStart={{ROOTLIBEXECDIR}}/systemd-sysroot-fstab-check
+Environment=SYSTEMD_INVOKED_AS=systemd-sysroot-fstab-check
+ExecStart={{SYSTEM_GENERATOR_DIR}}/systemd-fstab-generator
```
you are aware of the `@` syntax ExecStart= supports?

keszybz replied:
So yeah, for invocations in a unit this would work. But not for invocations on the command-line, which I think is useful too. But I don't have a strong view on this… Preferences?

keszybz replied:
I changed the unit to use `@`.

mrc0mmand commented on units/initrd-parse-etc.service.in:
```diff
@@ -0,0 +1,33 @@
+
+# FIXME: once dracut is patched to install the symlink, change to:
+# ExecStart={{ROOTLIBEXECDIR}}/systemd-sysroot-fstab-check
+ExecStart=@systemd-sysroot-fstab-check {{SYSTEM_GENERATOR_DIR}}/systemd-fstab-generator
```
Shouldn't this be the other way around?

```suggestion
ExecStart=@{{SYSTEM_GENERATOR_DIR}}/systemd-fstab-generator systemd-sysroot-fstab-check
```

> If the executable path is prefixed with "@", the second specified token will be passed as "argv[0]" to the executed process (instead of the actual filename), followed by the further arguments specified.
